### ISSUE TITLE
Calling double-through association on unsaved object returns nil

### ISSUE THREAD

**Issue by jacyreese** [rails/rails#20827]
Example of the issue:

``` ruby
class Response < ActiveRecord::Base
  belongs_to(
    :answer_choice,
    class_name: :AnswerChoice,
    primary_key: :id,
    foreign_key: :answer_choice_id
  )

  has_one :question, through: :answer_choice, source: :question
  has_one :poll, through: :question, source: :poll
end
```

``` ruby
class AnswerChoice < ActiveRecord::Base
  belongs_to(
    :question,
    class_name: :Question,
    primary_key: :id,
    foreign_key: :question_id
  )
end
```

``` ruby
class Question < ActiveRecord::Base
  belongs_to(
    :poll,
    class_name: :Poll,
    primary_key: :id,
    foreign_key: :poll_id
  )
end
```

``` ruby
class Poll < ActiveRecord::Base
end
```

In this example, when you create a new `response`, you can call `response.question` and get the corresponding question. When you try to call `response.poll` on an unsaved `response`, you get a return value of nil and no query is made to the database. Calling `response.question.poll` works fine, as does calling `response.question` on a `response` saved in the database. The issue seems to only come up when you call a double-through association on an unsaved object. For example, when you want to validate a `response` based on whether the `user` giving the response was the one who created the `poll` (so we wouldn't want them rigging their own poll!).

**Comment by rafaelfranca** [comment]
Which version are you using?

**Comment by jacyanthis** [comment]
@rafaelfranca I'm using 4.2.1 rails.

**Comment by rafaelfranca** [comment]
Could you create a reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

**Comment by jacyanthis** [comment]
Thanks for the reference. I made such a script here (https://github.com/jacyanthis/rails-bug-double-through/blob/master/active_record_bug_double_through.rb) although I can't get it to run. It just stalls when it tries to fetch rails from Github.

**Comment by rafaelfranca** [comment]
Thank you

**Comment by meinac** [comment]
I've created a pr for fixing this.

**Comment by al2o3cr** [comment]
Has this ever worked? It's unclear to me if this is even something doable in the general case, since the associations involved could be scoped.

In this particular case, I'd recommend bringing the reference to the user into the `Question` object, so something like:
```ruby
class Response < ActiveRecord::Base
  belongs_to :answer_choice
  belongs_to :response_by, class_name: 'User'

  has_one :question, through: :answer_choice
  has_one :poll, through: :question

  validate :check_for_ballot_stuffing

  def check_for_ballot_stuffing
    errors.add(:base, "stop stuffing ballots") if response_by == question.created_by
  end
end

class AnswerChoice < ActiveRecord::Base
  belongs_to :question
end

class Question < ActiveRecord::Base
  belongs_to :poll
  has_one :created_by, through: :poll
end

class Poll < ActiveRecord::Base
  belongs_to :created_by, class_name: 'User'
end
```
(I've elided some of the unneeded options from the original associations, and assumed you have a `User` class)

**Comment by meinac** [comment]
@al2o3cr the problem is that normally you can access post association of the subreply object but in case of new subreply(not persisted) you can't because of this bug. 
Try to save subreply object and reload it then you can access post association of it.

**Comment by wrkrb33** [comment]
I am running into this bug under 4.2.7.1, and have confirmed that it still exists in 5.0.0.1 as well.  Any chance that @meinac's pull request might get merged in someday?

**Comment by beydogan** [comment]
I can confirm that same bug still exists in 5.0.6

**Comment by kamipo** [comment]
Fixed by #16314.

**Comment by jherdman** [comment]
I'm running into this as recent as of Rails 5.2.3.
